core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala

```diff
@@ -168,6 +169,8 @@ private[spark] class TaskSetManager(
     t.epoch = epoch
   }
 
+  private val sortedPendingTasks = new AtomicBoolean(false)
```

Comment:
I think all of this stuff w/ the delayed sorting etc. is now totally unnecessary.  If you've got input sizes, then just sort the tasks when the TSM is created, which avoids a lot of complexity.

Perhaps the tasks should even just get sorted when the TaskSet is created in the first place, and then this doesn't know or care that the tasks have been sorted in any particular way.